### ISSUE TITLE
[java] Null assignment with developer intent

### ISSUE THREAD

**Issue by 0pencoding** [pmd/pmd#2455]
**Affects PMD Version:** 
6.20

**Rule:**
NullAssignment
<https://pmd.github.io/latest/pmd_rules_java_errorprone.html#nullassignment>

**Description:**
For the case where variable 'next' is not null, the developer intentionally makes 'next' into null. In this case, detecting this as an error is false positive.

**Code Sample demonstrating the issue:**

```java
private WindowedValue<T> consumeCurrent(){
  if (next == null) {
    throw new NoSuchElementException();
  }
 else {
    final WindowedValue<T> current=next;
    next=null;
    return current;
  }
}
```

**Expected outcome:**
*   Does PMD report a violation, where there shouldn't be one? -> false-positive

**Running PMD through:** *[CLI | Ant | Maven | Gradle | Designer | Other]*
CLI

**Comment by jsotuyod** [comment]
@0pencoding thanks for the report.

I'm unsure I understand your point here. The rule detects null assignments outside of declarations to avoid NPEs / changing values during execution (ie: if you already null-tested a var, it should be safe to access that var). You are assigning a variable to null outside a declaration, so the rule flags it. This is the expected behavior.

I understand in this particular case it may be your intention to do this on purpose, but I can't see how this should be evident to the rule itself. How is the rule supposed to tell "this null assignment is ok, this one isn't"?

It looks like you either want to suppress the rule in this particular statement (possible, check the docs on [rule suppression](https://pmd.github.io/pmd-6.23.0/pmd_userdocs_suppressing_warnings.html)), or simply want to exclude this rule from your project altogether.

**Comment by 0pencoding** [comment]
@jsotuyod  Thanks for the comment :)

I apologize in advance for my poor English,

We are doing research to reduce false positives of static analysis tools, and false positives are determined whether it was recognized as a bug or not from the developer's point of view. (This criterion is applicable to our research.)

Therefore, since the above variable 'next' was confirmed whether it was null or not and the null was assigned if it was not null, the developer believed that his intention would have been put in and that it would be determined as False Positive from the perspective of our research criteria mentioned earlier.

However, as you commented, given the purpose of the Null assignment rule itself, our point of view may be a little off.

Thank you for your sincere comment and creating a wonderful tool, PMD.
